File: src/main/java/org/moinex/ui/common/ResumePaneController.java
/*
 * Filename: ResumePaneController.java
 * Created on: October 10, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.common;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.Year;
import java.time.YearMonth;
import java.util.List;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import lombok.NoArgsConstructor;
import org.moinex.model.enums.WalletTransactionStatus;
import org.moinex.model.enums.WalletTransactionType;
import org.moinex.model.wallettransaction.WalletTransaction;
import org.moinex.service.PreferencesService;
import org.moinex.service.creditcard.CreditCardService;
import org.moinex.service.wallet.RecurringTransactionService;
import org.moinex.service.wallet.WalletService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Controller;

/**
 * Controller for the resume pane
 *
 * @note prototype is necessary so that each scene has its own controller
 */
@Controller
@Scope("prototype") // Each instance of this controller is unique
@NoArgsConstructor
public class ResumePaneController {
    @FXML private Label incomesCurrentSign;

    @FXML private Label incomesCurrentValue;

    @FXML private Label incomesForeseenSign;

    @FXML private Label incomesForeseenValue;

    @FXML private Label expensesCurrentSign;

    @FXML private Label expensesCurrentValue;

    @FXML private Label expensesForeseenSign;

    @FXML private Label expensesForeseenValue;

    @FXML private Label balanceCurrentSign;

    @FXML private Label balanceCurrentValue;

    @FXML private Label balanceForeseenSign;

    @FXML private Label balanceForeseenValue;

    @FXML private Label savingsCurrentSign;

    @FXML private Label savingsCurrentValue;

    @FXML private Label savingsForeseenSign;

    @FXML private Label savingsForeseenValue;

    @FXML private Label savingsLabel;

    @FXML private Label creditCardsCurrentSign;

    @FXML private Label creditCardsCurrentValue;

    @FXML private Label creditCardsForeseenSign;

    @FXML private Label creditCardsForeseenValue;

    private WalletService walletService;

    private RecurringTransactionService recurringTransactionService;

    private CreditCardService creditCardService;
    private PreferencesService preferencesService;

    /**
     * Constructor
     *
     * @param walletService WalletService
     * @param recurringTransactionService RecurringTransactionService
     * @param creditCardService CreditCardService
     * @note This constructor is used for dependency injection
     */
    @Autowired
    public ResumePaneController(
            WalletService walletService,
            RecurringTransactionService recurringTransactionService,
            CreditCardService creditCardService,
            PreferencesService preferencesService) {
        this.walletService = walletService;
        this.recurringTransactionService = recurringTransactionService;
        this.creditCardService = creditCardService;
        this.preferencesService = preferencesService;
    }

    @FXML
    public void initialize() {
        LocalDateTime now = LocalDateTime.now();
        updateResumePane(now.getMonthValue(), now.getYear());
    }

    /** Update the display of the resume */
    public void updateResumePane(Integer year) {
        List<WalletTransaction> allYearTransactions =
                walletService.getAllNonArchivedWalletTransactionsByYear(Year.of(year));

        List<WalletTransaction> futureTransactions =
                recurringTransactionService.getFutureRecurringTransactionsByYear(
                        Year.of(year), Year.of(year));

        allYearTransactions.addAll(futureTransactions);

        BigDecimal crcTotalDebtAmount = creditCardService.getTotalDebtAmountByYear(Year.of(year));

        BigDecimal crcPendingPayments =
                creditCardService.getTotalPendingPaymentsByYear(Year.of(year));

        BigDecimal crcPaidPayments = creditCardService.getTotalPaidPaymentsByYear(Year.of(year));

        updateResumePane(
                allYearTransactions, crcTotalDebtAmount, crcPendingPayments, crcPaidPayments);
    }

    /** Update the display of the month resume */
    public void updateResumePane(Integer month, Integer year) {
        // Get all transactions of the month that should be included in analysis
        List<WalletTransaction> transactions =
                walletService.getAllNonArchivedWalletTransactionsByMonthForAnalysis(
                        YearMonth.of(year, month));

        List<WalletTransaction> futureTransactions =
                recurringTransactionService.getFutureRecurringTransactionsByMonthForAnalysis(
                        YearMonth.of(year, month), YearMonth.of(year, month));

        transactions.addAll(futureTransactions);

        BigDecimal crcTotalDebtAmount =
                creditCardService.getTotalDebtAmountByMonth(YearMonth.of(year, month));

        BigDecimal crcPendingPayments =
                creditCardService.getTotalPendingPaymentsByMonth(YearMonth.of(year, month));

        BigDecimal crcPaidPayments =
                creditCardService.getTotalEffectivePaidPaymentsByMonth(YearMonth.of(year, month));

        updateResumePane(transactions, crcTotalDebtAmount, crcPendingPayments, crcPaidPayments);
    }

    private void updateResumePane(
            List<WalletTransaction> transactions,
            BigDecimal crcTotalDebtAmount,
            BigDecimal crcTotalPendingPayments,
            BigDecimal crcTotalPaidPayments) {
        BigDecimal totalConfirmedIncome =
                transactions.stream()
                        .filter(t -> t.getType().equals(WalletTransactionType.INCOME))
                        .filter(t -> t.getStatus().equals(WalletTransactionStatus.CONFIRMED))
                        .map(WalletTransaction::getAmount)
                        .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal totalConfirmedExpenses =
                transactions.stream()
                        .filter(t -> t.getType().equals(WalletTransactionType.EXPENSE))
                        .filter(t -> t.getStatus().equals(WalletTransactionStatus.CONFIRMED))
                        .map(WalletTransaction::getAmount)
                        .reduce(BigDecimal.ZERO, BigDecimal::add);

        // Consider the paid payments of the credit card as total expenses
        totalConfirmedExpenses = totalConfirmedExpenses.add(crcTotalPaidPayments);

        BigDecimal totalForeseenIncome =
                transactions.stream()
                        .filter(t -> t.getType() == WalletTransactionType.INCOME)
                        .map(WalletTransaction::getAmount)
                        .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal totalForeseenExpenses =
                transactions.stream()
                        .filter(t -> t.getType() == WalletTransactionType.EXPENSE)
                        .map(WalletTransaction::getAmount)
                        .reduce(BigDecimal.ZERO, BigDecimal::add);

        // Consider the payments of the credit card as total of foreseen expenses
        totalForeseenExpenses =
                totalForeseenExpenses.add(crcTotalPendingPayments).add(crcTotalPaidPayments);

        BigDecimal balance = totalConfirmedIncome.subtract(totalConfirmedExpenses);

        incomesCurrentValue.setText(UIUtils.formatCurrency(totalConfirmedIncome));
        incomesCurrentSign.setText(" "); // default
        incomesCurrentValue.getStyleClass().clear();
        incomesCurrentValue.getStyleClass().add(Constants.POSITIVE_BALANCE_STYLE);

        incomesForeseenValue.setText(UIUtils.formatCurrency(totalForeseenIncome));
        incomesForeseenSign.setText(" "); // default

        // Total Expenses
        expensesCurrentValue.setText(UIUtils.formatCurrency(totalConfirmedExpenses));
        expensesCurrentSign.setText(" "); // default
        expensesCurrentValue.getStyleClass().clear();
        expensesCurrentValue.getStyleClass().add(Constants.NEGATIVE_BALANCE_STYLE);

        expensesForeseenValue.setText(UIUtils.formatCurrency(totalForeseenExpenses));
        expensesForeseenSign.setText(" "); // default

        // Balance
        balanceCurrentValue.setText(UIUtils.formatCurrency(balance));

        // Set the balance label and sign label according to the balance value
        if (balance.compareTo(BigDecimal.ZERO) > 0) {
            balanceCurrentValue.setText(UIUtils.formatCurrency(balance));
            balanceCurrentSign.setText("+");

            balanceCurrentValue.getStyleClass().clear();
            balanceCurrentValue.getStyleClass().add(Constants.POSITIVE_BALANCE_STYLE);

            balanceCurrentSign.getStyleClass().clear();
            balanceCurrentSign.getStyleClass().add(Constants.POSITIVE_BALANCE_STYLE);
        } else if (balance.compareTo(BigDecimal.ZERO) < 0) {
            balanceCurrentValue.setText(UIUtils.formatCurrency(balance.abs()));
            balanceCurrentSign.setText("-");

            balanceCurrentValue.getStyleClass().clear();
            balanceCurrentValue.getStyleClass().add(Constants.NEGATIVE_BALANCE_STYLE);

            balanceCurrentSign.getStyleClass().clear();
            balanceCurrentSign.getStyleClass().add(Constants.NEGATIVE_BALANCE_STYLE);
        } else {
            balanceCurrentValue.setText(UIUtils.formatCurrency(0.0));
            balanceCurrentSign.setText("");

            balanceCurrentValue.getStyleClass().clear();
            balanceCurrentValue.getStyleClass().add(Constants.NEUTRAL_BALANCE_STYLE);

            balanceCurrentSign.getStyleClass().clear();
            balanceCurrentSign.getStyleClass().add(Constants.NEUTRAL_BALANCE_STYLE);
        }

        BigDecimal foreseenBalance = totalForeseenIncome.subtract(totalForeseenExpenses);

        if (foreseenBalance.compareTo(BigDecimal.ZERO) > 0) {
            balanceForeseenValue.setText(UIUtils.formatCurrency(foreseenBalance));
            balanceForeseenSign.setText("+");
        } else if (foreseenBalance.compareTo(BigDecimal.ZERO) < 0) {
            balanceForeseenValue.setText(UIUtils.formatCurrency(foreseenBalance.abs()));
            balanceForeseenSign.setText("-");
        } else {
            balanceForeseenValue.setText(UIUtils.formatCurrency(0.0));
            balanceForeseenSign.setText(" ");
        }

        // Mensal Economies
        double savingsPercentage;

        if (totalConfirmedIncome.compareTo(BigDecimal.ZERO) <= 0) {
            savingsPercentage = 0.0;
        } else {
            savingsPercentage =
                    totalConfirmedIncome.subtract(totalConfirmedExpenses).doubleValue()
                            / totalConfirmedIncome.doubleValue()
                            * 100;
        }

        // Set the economy label and sign label according to the economy value
        UIUtils.removeTooltipFromNode(savingsCurrentValue);

        if (savingsPercentage > 0) {
            savingsLabel.setText(
                    preferencesService.translate(Constants.TranslationKeys.COMMON_RESUME_SAVINGS));
            savingsCurrentValue.setText(
                    UIUtils.formatPercentage(savingsPercentage, preferencesService));
            savingsCurrentSign.setText("+");

            savingsCurrentValue.getStyleClass().clear();
            savingsCurrentValue.getStyleClass().add(Constants.POSITIVE_BALANCE_STYLE);

            savingsCurrentSign.getStyleClass().clear();
            savingsCurrentSign.getStyleClass().add(Constants.POSITIVE_BALANCE_STYLE);
        } else if (savingsPercentage < 0) {
            savingsLabel.setText(
                    preferencesService.translate(
                            Constants.TranslationKeys.COMMON_RESUME_NO_SAVINGS));
            savingsCurrentValue.setText(
                    UIUtils.formatPercentage(savingsPercentage, preferencesService));

            if (savingsPercentage < Constants.NEGATIVE_PERCENTAGE_THRESHOLD) {
                savingsCurrentSign.setText(" ");
                UIUtils.addTooltipToNode(
                        savingsCurrentValue,
                        "- " + UIUtils.formatPercentage(-savingsPercentage, preferencesService));
            } else {
                savingsCurrentSign.setText("-");
            }

            savingsCurrentValue.getStyleClass().clear();
            savingsCurrentValue.getStyleClass().add(Constants.NEGATIVE_BALANCE_STYLE);

            savingsCurrentSign.getStyleClass().clear();
            savingsCurrentSign.getStyleClass().add(Constants.NEGATIVE_BALANCE_STYLE);
        } else {
            savingsLabel.setText(
                    preferencesService.translate(
                            Constants.TranslationKeys.COMMON_RESUME_NO_SAVINGS));
            savingsCurrentValue.setText(UIUtils.formatPercentage(0.0, preferencesService));
            savingsCurrentSign.setText(" ");

            savingsCurrentValue.getStyleClass().clear();
            savingsCurrentValue.getStyleClass().add(Constants.NEUTRAL_BALANCE_STYLE);

            savingsCurrentSign.getStyleClass().clear();
            savingsCurrentSign.getStyleClass().add(Constants.NEUTRAL_BALANCE_STYLE);
        }

        double foreseenSavingsPercentage = 0.0;

        if (totalForeseenIncome.compareTo(BigDecimal.ZERO) > 0) {
            foreseenSavingsPercentage =
                    totalForeseenIncome.subtract(totalForeseenExpenses).doubleValue()
                            / totalForeseenIncome.doubleValue()
                            * 100;
        }

        if (foreseenSavingsPercentage > 0) {
            savingsForeseenValue.setText(
                    UIUtils.formatPercentage(foreseenSavingsPercentage, preferencesService));
            savingsForeseenSign.setText("+");
        } else if (foreseenSavingsPercentage < 0) {
            savingsForeseenValue.setText(
                    UIUtils.formatPercentage(-foreseenSavingsPercentage, preferencesService));
            savingsForeseenSign.setText("-");
        } else {
            savingsForeseenValue.setText(UIUtils.formatPercentage(0.0, preferencesService));
            savingsForeseenSign.setText(" ");
        }

        // Credit Card
        creditCardsCurrentValue.setText(UIUtils.formatCurrency(crcTotalDebtAmount));
        creditCardsCurrentSign.setText(" "); // default

        creditCardsForeseenValue.setText(
                String.format(UIUtils.formatCurrency(crcTotalPendingPayments)));
        creditCardsForeseenSign.setText(" "); // default
    }
}

File: src/main/kotlin/org/moinex/ui/common/ResumePaneController.kt
/*
 * Filename: ResumePaneController.kt (original filename: ResumePaneController.java)
 * Created on: October 10, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 19/03/2026
 */

package org.moinex.ui.common

import javafx.fxml.FXML
import javafx.scene.control.Label
import org.moinex.common.constant.Constants
import org.moinex.common.constant.Styles
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.extension.isConfirmed
import org.moinex.common.extension.isExpense
import org.moinex.common.extension.isIncome
import org.moinex.common.util.UIUtils
import org.moinex.model.wallettransaction.WalletTransaction
import org.moinex.service.PreferencesService
import org.moinex.service.creditcard.CreditCardService
import org.moinex.service.creditcard.RecurringCreditCardDebtService
import org.moinex.service.wallet.RecurringTransactionService
import org.moinex.service.wallet.WalletService
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.math.RoundingMode
import java.time.LocalDateTime
import java.time.Year
import java.time.YearMonth

@Controller
@Scope("prototype")
class ResumePaneController(
    private val walletService: WalletService,
    private val recurringTransactionService: RecurringTransactionService,
    private val creditCardService: CreditCardService,
    private val recurringCreditCardDebtService: RecurringCreditCardDebtService,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var incomesCurrentSign: Label

    @FXML
    private lateinit var incomesCurrentValue: Label

    @FXML
    private lateinit var incomesForeseenSign: Label

    @FXML
    private lateinit var incomesForeseenValue: Label

    @FXML
    private lateinit var expensesCurrentSign: Label

    @FXML
    private lateinit var expensesCurrentValue: Label

    @FXML
    private lateinit var expensesForeseenSign: Label

    @FXML
    private lateinit var expensesForeseenValue: Label

    @FXML
    private lateinit var balanceCurrentSign: Label

    @FXML
    private lateinit var balanceCurrentValue: Label

    @FXML
    private lateinit var balanceForeseenSign: Label

    @FXML
    private lateinit var balanceForeseenValue: Label

    @FXML
    private lateinit var savingsCurrentSign: Label

    @FXML
    private lateinit var savingsCurrentValue: Label

    @FXML
    private lateinit var savingsForeseenSign: Label

    @FXML
    private lateinit var savingsForeseenValue: Label

    @FXML
    private lateinit var savingsLabel: Label

    @FXML
    private lateinit var creditCardsCurrentSign: Label

    @FXML
    private lateinit var creditCardsCurrentValue: Label

    @FXML
    private lateinit var creditCardsForeseenSign: Label

    @FXML
    private lateinit var creditCardsForeseenValue: Label

    companion object {
        private const val DEFAULT_SIGN = " "
        private const val POSITIVE_SIGN = "+"
        private const val NEGATIVE_SIGN = "-"
        private const val EMPTY_SIGN = ""
        private val PERCENTAGE_MULTIPLIER = BigDecimal(100)
    }

    @FXML
    fun initialize() {
        val now = LocalDateTime.now()
        updateResumePane(Year.from(now))
    }

    fun updateResumePane(year: Year) {
        val allYearTransactions =
            walletService.getAllNonArchivedWalletTransactionsByYear(year)

        val futureTransactions =
            recurringTransactionService.getFutureRecurringTransactionsByYear(
                year,
                year,
            )

        val allTransactions = allYearTransactions + futureTransactions

        val projectedCrcAmount = recurringCreditCardDebtService.getTotalProjectedAmountForYear(year)

        val crcTotalDebtAmount = creditCardService.getTotalDebtAmountByYear(year) + projectedCrcAmount
        val crcPendingPayments = creditCardService.getTotalPendingPaymentsByYear(year) + projectedCrcAmount
        val crcPaidPayments = creditCardService.getTotalPaidPaymentsByYear(year)

        updateResumePane(allTransactions, crcTotalDebtAmount, crcPendingPayments, crcPaidPayments)
    }

    fun updateResumePane(yearMonth: YearMonth) {
        val transactions =
            walletService.getAllNonArchivedWalletTransactionsByMonthForAnalysis(yearMonth)

        val futureTransactions =
            recurringTransactionService.getFutureRecurringTransactionsByMonthForAnalysis(
                yearMonth,
                yearMonth,
            )

        val allTransactions = transactions + futureTransactions

        val projectedCrcAmount = recurringCreditCardDebtService.getTotalProjectedAmountForMonth(yearMonth)

        val crcTotalDebtAmount = creditCardService.getTotalDebtAmountByMonth(yearMonth) + projectedCrcAmount
        val crcPendingPayments = creditCardService.getTotalPendingPaymentsByMonth(yearMonth) + projectedCrcAmount
        val crcPaidPayments = creditCardService.getTotalEffectivePaidPaymentsByMonth(yearMonth)

        updateResumePane(allTransactions, crcTotalDebtAmount, crcPendingPayments, crcPaidPayments)
    }

    private fun updateResumePane(
        transactions: List<WalletTransaction>,
        crcTotalDebtAmount: BigDecimal,
        crcTotalPendingPayments: BigDecimal,
        crcTotalPaidPayments: BigDecimal,
    ) {
        val totalConfirmedIncome =
            transactions
                .filter { it.isIncome() }
                .filter { it.isConfirmed() }
                .sumOf { it.amount }

        val totalConfirmedExpenses =
            transactions
                .filter { it.isExpense() }
                .filter { it.isConfirmed() }
                .sumOf { it.amount } + crcTotalPaidPayments

        val totalForeseenIncome =
            transactions
                .filter { it.isIncome() }
                .sumOf { it.amount }

        val totalForeseenExpenses =
            transactions
                .filter { it.isExpense() }
                .sumOf { it.amount } + crcTotalPendingPayments + crcTotalPaidPayments

        val balance = totalConfirmedIncome - totalConfirmedExpenses
        val foreseenBalance = totalForeseenIncome - totalForeseenExpenses

        updateIncomeLabels(totalConfirmedIncome, totalForeseenIncome)
        updateExpenseLabels(totalConfirmedExpenses, totalForeseenExpenses)
        updateBalanceLabels(balance, foreseenBalance)
        updateSavingsLabels(totalConfirmedIncome, totalConfirmedExpenses, totalForeseenIncome, totalForeseenExpenses)
        updateCreditCardLabels(crcTotalDebtAmount, crcTotalPendingPayments)
    }

    private fun updateIncomeLabels(
        confirmed: BigDecimal,
        foreseen: BigDecimal,
    ) {
        incomesCurrentValue.apply {
            text = UIUtils.formatCurrency(confirmed)
            styleClass.apply {
                clear()
                add(Styles.POSITIVE_BALANCE_STYLE)
            }
        }
        incomesCurrentSign.text = DEFAULT_SIGN

        incomesForeseenValue.text = UIUtils.formatCurrency(foreseen)
        incomesForeseenSign.text = DEFAULT_SIGN
    }

    private fun updateExpenseLabels(
        confirmed: BigDecimal,
        foreseen: BigDecimal,
    ) {
        expensesCurrentValue.apply {
            text = UIUtils.formatCurrency(confirmed)
            styleClass.apply {
                clear()
                add(Styles.NEGATIVE_BALANCE_STYLE)
            }
        }
        expensesCurrentSign.text = DEFAULT_SIGN

        expensesForeseenValue.text = UIUtils.formatCurrency(foreseen)
        expensesForeseenSign.text = DEFAULT_SIGN
    }

    private fun updateBalanceLabels(
        current: BigDecimal,
        foreseen: BigDecimal,
    ) {
        when {
            current > BigDecimal.ZERO -> {
                balanceCurrentValue.text = UIUtils.formatCurrency(current)
                balanceCurrentSign.text = POSITIVE_SIGN
                applyStyleToLabels(
                    listOf(balanceCurrentValue, balanceCurrentSign),
                    Styles.POSITIVE_BALANCE_STYLE,
                )
            }
            current < BigDecimal.ZERO -> {
                balanceCurrentValue.text = UIUtils.formatCurrency(current.abs())
                balanceCurrentSign.text = NEGATIVE_SIGN
                applyStyleToLabels(
                    listOf(balanceCurrentValue, balanceCurrentSign),
                    Styles.NEGATIVE_BALANCE_STYLE,
                )
            }
            else -> {
                balanceCurrentValue.text = UIUtils.formatCurrency(BigDecimal.ZERO)
                balanceCurrentSign.text = EMPTY_SIGN
                applyStyleToLabels(
                    listOf(balanceCurrentValue, balanceCurrentSign),
                    Styles.NEUTRAL_BALANCE_STYLE,
                )
            }
        }

        when {
            foreseen > BigDecimal.ZERO -> {
                balanceForeseenValue.text = UIUtils.formatCurrency(foreseen)
                balanceForeseenSign.text = POSITIVE_SIGN
            }
            foreseen < BigDecimal.ZERO -> {
                balanceForeseenValue.text = UIUtils.formatCurrency(foreseen.abs())
                balanceForeseenSign.text = NEGATIVE_SIGN
            }
            else -> {
                balanceForeseenValue.text = UIUtils.formatCurrency(BigDecimal.ZERO)
                balanceForeseenSign.text = DEFAULT_SIGN
            }
        }
    }

    private fun updateSavingsLabels(
        confirmedIncome: BigDecimal,
        confirmedExpenses: BigDecimal,
        foreseenIncome: BigDecimal,
        foreseenExpenses: BigDecimal,
    ) {
        val savingsPercentage =
            if (confirmedIncome <= BigDecimal.ZERO) {
                BigDecimal.ZERO
            } else {
                (
                    confirmedIncome.minus(
                        confirmedExpenses,
                    )
                ).divide(confirmedIncome, 2, RoundingMode.HALF_UP).multiply(PERCENTAGE_MULTIPLIER)
            }

        UIUtils.removeTooltipFromNode(savingsCurrentValue)

        when {
            savingsPercentage > BigDecimal.ZERO -> {
                savingsLabel.text = preferencesService.translate(TranslationKeys.COMMON_RESUME_SAVINGS)
                savingsCurrentValue.text = UIUtils.formatPercentage(savingsPercentage)
                savingsCurrentSign.text = POSITIVE_SIGN
                applyStyleToLabels(
                    listOf(savingsCurrentValue, savingsCurrentSign),
                    Styles.POSITIVE_BALANCE_STYLE,
                )
            }
            savingsPercentage < BigDecimal.ZERO -> {
                savingsLabel.text = preferencesService.translate(TranslationKeys.COMMON_RESUME_NO_SAVINGS)
                savingsCurrentValue.text = UIUtils.formatPercentage(savingsPercentage)

                if (savingsPercentage < Constants.NEGATIVE_PERCENTAGE_THRESHOLD.toBigDecimal()) {
                    savingsCurrentSign.text = DEFAULT_SIGN
                    UIUtils.addTooltipToNode(
                        savingsCurrentValue,
                        "- ${UIUtils.formatPercentage(-savingsPercentage)}",
                    )
                } else {
                    savingsCurrentSign.text = NEGATIVE_SIGN
                }

                applyStyleToLabels(
                    listOf(savingsCurrentValue, savingsCurrentSign),
                    Styles.NEGATIVE_BALANCE_STYLE,
                )
            }
            else -> {
                savingsLabel.text = preferencesService.translate(TranslationKeys.COMMON_RESUME_NO_SAVINGS)
                savingsCurrentValue.text = UIUtils.formatPercentage(BigDecimal.ZERO)
                savingsCurrentSign.text = DEFAULT_SIGN
                applyStyleToLabels(
                    listOf(savingsCurrentValue, savingsCurrentSign),
                    Styles.NEUTRAL_BALANCE_STYLE,
                )
            }
        }

        val foreseenSavingsPercentage =
            if (foreseenIncome > BigDecimal.ZERO) {
                (
                    foreseenIncome.minus(
                        foreseenExpenses,
                    )
                ).divide(foreseenIncome, 2, RoundingMode.HALF_UP).multiply(PERCENTAGE_MULTIPLIER)
            } else {
                BigDecimal.ZERO
            }

        when {
            foreseenSavingsPercentage > BigDecimal.ZERO -> {
                savingsForeseenValue.text = UIUtils.formatPercentage(foreseenSavingsPercentage)
                savingsForeseenSign.text = POSITIVE_SIGN
            }
            foreseenSavingsPercentage < BigDecimal.ZERO -> {
                savingsForeseenValue.text = UIUtils.formatPercentage(-foreseenSavingsPercentage)
                savingsForeseenSign.text = NEGATIVE_SIGN
            }
            else -> {
                savingsForeseenValue.text = UIUtils.formatPercentage(BigDecimal.ZERO)
                savingsForeseenSign.text = DEFAULT_SIGN
            }
        }
    }

    private fun updateCreditCardLabels(
        totalDebt: BigDecimal,
        pendingPayments: BigDecimal,
    ) {
        creditCardsCurrentValue.text = UIUtils.formatCurrency(totalDebt)
        creditCardsCurrentSign.text = DEFAULT_SIGN

        creditCardsForeseenValue.text = UIUtils.formatCurrency(pendingPayments)
        creditCardsForeseenSign.text = DEFAULT_SIGN
    }

    private fun applyStyleToLabels(
        labels: List<Label>,
        styleClass: String,
    ) {
        labels.forEach { label ->
            label.styleClass.apply {
                clear()
                add(styleClass)
            }
        }
    }
}
